scripts/builtin/decisionTree.dml

```diff
@@ -0,0 +1,449 @@
+
+#
+# THIS SCRIPT IMPLEMENTS CLASSIFICATION TREES WITH BOTH SCALE AND CATEGORICAL FEATURES
+#
+# INPUT         PARAMETERS:
+# ---------------------------------------------------------------------------------------------
+# NAME          TYPE     DEFAULT      MEANING
+# ---------------------------------------------------------------------------------------------
+# X             String   ---          Location to read feature matrix X; note that X needs to be both recoded and dummy coded
+# Y             String   ---          Location to read label matrix Y; note that Y needs to be both recoded and dummy coded
+# R             String   " "          Location to read the matrix R which for each feature in X contains the following information
+#                                       - R[1,]: Row Vector which indicates if feature vector is scalar or categorical. 1 indicates
+#                                     a scalar feature vector, other positive Integers indicate the number of categories
+#                                     If R is not provided by default all variables are assumed to be scale
+# bins          Int      20           Number of equiheight bins per scale feature to choose thresholds
+# depth         Int      25           Maximum depth of the learned tree
+# M             String   ---          Location to write matrix M containing the learned tree
+# ---------------------------------------------------------------------------------------------
+# OUTPUT:
+# Matrix M where each column corresponds to a node in the learned tree and each row contains the following information:
+#    M[1,j]: id of node j (in a complete binary tree)
+#    M[2,j]: Offset (no. of columns) to left child of j if j is an internal node, otherwise 0
+#    M[3,j]: Feature index of the feature (scale feature id if the feature is scale or categorical feature id if the feature is categorical)
+#            that node j looks at if j is an internal node, otherwise 0
+#    M[4,j]: Type of the feature that node j looks at if j is an internal node: holds the same information as R input vector
+#    M[5,j]: If j is an internal node: 1 if the feature chosen for j is scale, otherwise the size of the subset of values
+#            stored in rows 6,7,... if j is categorical
+#            If j is a leaf node: number of misclassified samples reaching at node j
+#    M[6:,j]: If j is an internal node: Threshold the example's feature value is compared to is stored at M[6,j] if the feature chosen for j is scale,
+#             otherwise if the feature chosen for j is categorical rows 6,7,... depict the value subset chosen for j
+#             If j is a leaf node 1 if j is impure and the number of samples at j > threshold, otherwise 0
+# -------------------------------------------------------------------------------------------
+
+m_decisionTree = function(
+    Matrix[Double] X,
+    Matrix[Double] Y,
+    Matrix[Double] R,
+    int bins = 10,
+    int depth = 20
```

Comment:
add a verbose flag to enable and disable writing (you can find inspiration in other builtin scripts.)